Fix potential IP overlapping with node LB

xinfengliu opened the pull request:
Fixes #40989 

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
This PR removes node LB after container exits itself or starting failure to avoid IP overlapping.

**- How I did it**
Add `removeNetworks()` in `daemon/cluster/executor/container/controller.go`

**- How to verify it**
Manual verification using the test case in #40989
Also added test cases in CI in this PR.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Fix potential IP overlapping with node LB when container exits itself

**- A picture of a cute animal (not mandatory but encouraged)**

xinfengliu committed:
Remove node LB to avoid IP overlapping

Remove node LB after task container exits itself or startup failure.
For detailed reason, see https://github.com/moby/moby/issues/40989

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

thaJeztah commented:
@arkodg @dperny PTAL

xinfengliu commented:
I just think of another scenario of IP overlapping: 
Container starting failure (not because of networking)

Example:
`docker service create --network net0 --name bb1 --restart-max-attempts=1 --restart-delay=10s busybox non-existing-file` 

I will make some modification to the codes.

xinfengliu commented:
If we handle the network cleanup well in `Start()` and `Wait()`, then we don't have to remove networks in `Shutdown()` if the container is not running (either already stopped or just in `created` state).
 
This will avoid unnecessary removing networks in below swarmkit scenarios:
-  Desired task status is changed from READY to SHUTDOWN
-  Remove tasks in some situations, e.g. task history limit is reached.

And this will in turn help mitigate widely reported issue: `failed to get network during CreateEndpoint` (https://github.com/docker/for-linux/issues/888#issuecomment-628350989) .

xinfengliu commented:
Hi,
I feel it's ready for review. Thanks.
I have tried to consider various scenarios when I made code changes in the past 2 days and did manual tests to verify them.

mightydok commented:
Why we need remove whole network instead of just not give used for lb IP address to new containers?

thaJeztah commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -185,7 +185,7 @@ func (r *controller) Prepare(ctx context.Context) error {
 
 // Start the container. An error will be returned if the container is already started.
-func (r *controller) Start(ctx context.Context) error {
+func (r *controller) Start(ctx context.Context) (err error) {
```
Could you rename `err` to `retErr` to make it stand out as being this output-variable? `err` is used as name for local variables in this function, which makes it easy to confuse them (and to accidentally mask the output variable)

xinfengliu replied:
yes, I will rename it to make it clearer.

thaJeztah commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -217,17 +217,35 @@ func (r *controller) Start(ctx context.Context) error {
 	}
 
+	// At this time, the container is started via containerd, and container.State.Running is true
+	// However, there is still possible errors from now on in this Start() function, and swarmkit will not
+	// call Shutdown() after task state is FAIL. We must clean up here.
```
> and swarmkit will not call Shutdown() after task state is FAIL

Was this by design? Should this be changed? @dperny

xinfengliu replied:
This is how current swarmkit works:
https://github.com/docker/swarmkit/blob/master/agent/exec/controller.go#L295
```
	if task.DesiredState >= api.TaskStateShutdown {
		if status.State >= api.TaskStateCompleted {
			return noop()
		}
...
```

thaJeztah commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -281,7 +303,7 @@ func (r *controller) Start(ctx context.Context) error {
 
 // Wait on the container to exit.
-func (r *controller) Wait(pctx context.Context) error {
+func (r *controller) Wait(pctx context.Context) (err error) {
```
Same comment here (`s/err/retErr/`)

xinfengliu replied:
Done.

thaJeztah commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -370,6 +403,12 @@ func (r *controller) Shutdown(ctx context.Context) error {
 	}
 
 	if err := r.adapter.shutdown(ctx); err != nil {
+		// The container is already stopped or not even started.
+		// don't need to removeNetworks.
+		if isStoppedContainer(err) {
+			log.G(ctx).Debugf("Shutdown(): container %s already stopped.", r.adapter.container.name())
+			return nil
+		}
```
Looks like this was already handled by the line below;

```go
if !(isUnknownContainer(err) || isStoppedContainer(err)) {
```

So this is only adding a debug-log? Do we still need that, or did you only need it while working on your patch?

Also wondering; this looks to be the only place where `adapter.shutdown()` is called; should this logic be in that function?

xinfengliu replied:
Yes, I think we need debug logs, otherwise we don't know what happened on customer's production system.

thaJeztah commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -217,17 +217,35 @@ func (r *controller) Start(ctx context.Context) error {
+			// Since removeNeworks() is an idempotent operation, it should be OK to removeNetworks
+			// here although it's not ideal.
+			log.G(ctx).Debugf("Removing networks after container %s failed to start.", r.adapter.container.name())
+			r.adapter.removeNetworks(ctx)
```
Should errors that occur during removal of the network be handled? (perhaps only logged if we can't / don't have to handle them otherwise)

xinfengliu replied:
Yes, we need log the error of `removeNetworks(ctx)`, but return the error indicating the reason for container-start.

thaJeztah commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -530,3 +530,14 @@ func isUnknownContainer(err error) bool {
 }
+
+func isErrContainerUnhealthy(err error) bool {
+	if err != nil {
+		if e, ok := err.(*exitError); ok {
```
Should this use `errors.As()`? Something like;

```suggestion
	var eErr exitError
	if errors.As(err, & eErr) {
```

Perhaps use an early return for the `err == nil` case 🤔

xinfengliu replied:
Yes, will do.

xinfengliu replied:
Done.

thaJeztah commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -530,3 +530,14 @@ func isUnknownContainer(err error) bool {
+func isErrContainerUnhealthy(err error) bool {
+	if err != nil {
+		if e, ok := err.(*exitError); ok {
+			if c := e.Cause(); c != nil {
+				return strings.Contains(c.Error(), "unhealthy container")
+			}
```
We should probably avoid using `errors.Cause()` to prevent issues if something lower in the stack starts using native "go" wrapping of errors.

In this case, I'm wondering if just checking `strings.Contains()` on the `exitError` would suffice.

That said, if I see correctly, these errors are generated in https://github.com/moby/moby/blob/07d60bc2571ba3d680f21adc84d87803ab4959c6/daemon/cluster/executor/container/controller.go#L248-L273

Which is in this codebase, so could we instead wrap a specific error that we can check for here? (so modify the code I linked above, and make it wrap a typed error, so that we can check for that error-type instead of do string-matching?

xinfengliu replied:
Sorry, I don't quite understand what you mean. Here, i just want to filter out `exitError{code: ctnr.State.ExitCode, cause: healthErr} `. If I don't extract `cause` in `exitError`, I don't have a way to catch it.
Yes, It's better to check error-type instead of do string-matching for `healthErr`.  I will try doing it.

xinfengliu replied:
Modified the code per your suggestion.

xinfengliu commented:
@mightydok 
> Why we need remove whole network instead of just not give used for lb IP address to new containers?

We remove the network only when there's no containers running on this network on this node. The logic is in `delete()` in `vendor/github.com/docker/libnetwork/network.go`
If there's no container running on the network on the worker node, then this node should not be a peer node for this overlay network. So we need to remove the network on this node.

cpuguy83 commented on daemon/cluster/executor/container/adapter.go:
```diff
@@ -530,3 +530,14 @@ func isUnknownContainer(err error) bool {
 }
+
+func isErrContainerUnhealthy(err error) bool {
+	if err != nil {
```
```go
if err == nil {
    return false
}
```

xinfengliu replied:
Done.

cpuguy83 commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -185,7 +185,7 @@ func (r *controller) Prepare(ctx context.Context) error {
 
 // Start the container. An error will be returned if the container is already started.
-func (r *controller) Start(ctx context.Context) error {
+func (r *controller) Start(ctx context.Context) (err error) {
```
How about `retErr` instead of `err`.

xinfengliu replied:
Done.

cpuguy83 commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -217,17 +217,35 @@ func (r *controller) Start(ctx context.Context) error {
+		// Shutdown() has been called.
+		if err != nil && !isErrContainerUnhealthy(err) {
+			log.G(ctx).Debugf("Removing networks after container %s started but then failed.", r.adapter.container.name())
+			r.adapter.removeNetworks(ctx)
```
log errors

cpuguy83 commented on daemon/cluster/executor/container/controller.go:
```diff
@@ -217,17 +217,35 @@ func (r *controller) Start(ctx context.Context) error {
 	if ctnr.Config == nil || ctnr.Config.Healthcheck == nil || len(ctnr.Config.Healthcheck.Test) == 0 || ctnr.Config.Healthcheck.Test[0] == "NONE" {
 		if err := r.adapter.activateServiceBinding(); err != nil {
 			log.G(ctx).WithError(err).Errorf("failed to activate service binding for container %s which has no healthcheck config", r.adapter.container.name())
+			r.adapter.shutdown(ctx)
```
log errors

xinfengliu replied:
Done.

xinfengliu committed:
add tests for verifying node LB removal

add test cases for verifying node LB removal under some conditions.
Also add:
- `ServiceWithRestartAttempts` and `ServiceWithHealthCheck` for service creation in tests.
- Waiting conditions for `CompletedTasksCount` and `FailedTasksCount`

Signed-off-by: Xinfeng Liu <xinfeng.liu@gmail.com>

xinfengliu commented:
Added a new commit for tests:

- 	test NormalService
- 	test StartServiceFail
- 	test ServiceCompletion
- 	test ServiceExitFail
- 	test ServiceHealthCheckFail
- 	test ServiceHealthCheckExitEarly

xinfengliu commented:
Hi @thaJeztah 
I created another PR https://github.com/moby/moby/pull/41062 using a different way to address the same issue. I believe that way is much cleaner. Wish you to take a look. Thanks.
(Test cases are added to CI too)

thaJeztah commented:
Looks like this was addressed in https://github.com/moby/moby/pull/42432
